test/tesla/adapter/gun_test.exs

```diff
@@ -0,0 +1,50 @@
+defmodule Tesla.Adapter.GunTest do
+  use ExUnit.Case
+
+  use Tesla.AdapterCase, adapter: Tesla.Adapter.Gun
+  use Tesla.AdapterCase.Basic
+  use Tesla.AdapterCase.Multipart
+  use Tesla.AdapterCase.StreamRequestBody
+  use Tesla.AdapterCase.SSL
+
+  test "timeout option" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}/delay/2"
+    }
+
+    assert {:error, "read response timeout"} = Tesla.Adapter.Gun.call(request, timeout: 1_000)
+  end
```

Comment:
While there are some issues with Tesla.Middleware.Timeout I don't think we should expose such custom options as part of adapter interface. In general adapter options should be used as last-resort case when one absolutely have to pass some options to the underlying adapter.

Reply:
now it uses `opts[:timeout]` option or adapter default fallback setting